src/librustc_mir/transform/validate.rs

```diff
@@ -244,6 +252,35 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
                             );
                         }
                     }
+                    // The deaggregator currently does not deaggreagate arrays.
+                    // Until then, we ignore them here.
+                    Rvalue::Aggregate(box AggregateKind::Array { .. }, _) => {}
+                    // All other aggregates must be gone after some phases.
+                    Rvalue::Aggregate(box kind, _) => {
+                        match self.mir_phase {
+                            MirPhase::Build
+                            | MirPhase::Const
+                            | MirPhase::ConstPromotion
+                            // Until the drop elaboration phase finishes, we still have aggregates.
+                            | MirPhase::DropLowering => {}
+                            MirPhase::GeneratorLowering => {
+                                // Generators persist until the state machine transformation.
```

Comment:
I wonder if this would become more readable by also using the `>` scheme.
```rust
// Only generators are permitted after drop lowering.
if mir_phase > DropLowering && !matches!(kind, AggregateKind::Generator(..)) {
  self.fail(...)
}
// All aggregates are gone after generator lowering.
if mir_phase > GeneratorLowering {
  self.fail(...)
}
```

Reply:
I tried that first, but the `else if` had me run in circles in my head trying to understand what happens where.

Reply:
Hm, I find that much better than the `match` though... the first branch has a `fail` so the `else` is easy to justify IMO.

Alternatively we could `return` after the first `fail`, and we do not need the `else`.

Reply:
I refactored to using `>` and no match, it's different from what you suggested, but felt clearer to me

Reply:
Uff, a negation below a disjunction... that took a while to make sure it makes sense. IMO the if-chain is a *lot* easier to understand.

Not necessarily a blocker, but funny how we cannot seem to find a way to write this code that's readable for both of us.^^

Reply:
Nested ifs are equivalent to a conjunction, so this is the same as
```rust
mir_phase > MirPhase::DropLowering && (mir_phase > MirPhase::GeneratorLowering || !matches(...))`
```
and I find them also both equal (aka very low) in readability.

Reply:
In contrast, the code I proposed directly encodes the two invariants that we want to enforce:
* after drop elab, no aggregates except generators
* after generator elab, no aggregates

Those invariants are hart to tease out of your code, I feel.

Reply:
yea, I've come around to your solution. I wish we could have range patterns for enums with explicit discriminants. Then we could just write all of this as patterns